### ISSUE TITLE
Implement a `@must_await` annotation for functions

### ISSUE THREAD

**Issue by IntangibleMatter** [godotengine/godot-proposals#11094]
### Describe the project you are working on

An RPG with a lot of cutscenes

### Describe the problem or limitation you are having in your project

For a variety of reasons, I've elected to use GDScript to write the cutscenes for the game. It works well in most cases, but there's a key problem: needing to constantly insert `await` before function calls.

Take this sample script:
```gdscript
func cutscene_label() -> void:
    player.position = villain.position + Vector2(0, 128)
    await say("player", "ENDING_ENCOUNTER_0")
    await say("player", "ENDING_ENCOUNTER_1")
    villain.animate("eyeflame")
    await say("villain", "ENDING_ENCOUNTER_2")
```
When the call to a function should be awaited by default, it gets irritatingly verbose to put this ahead of it every time.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Implement a function annotation along the lines of `@await_by_default`, which would make it so that any calls to that function implicitly have an `await`, instead of needing to insert it manually. This would allow for decreasing the verbosity of code that expects several actions to run in sequence despite taking multiple frames, allowing GDScript to act better as a scripting language for some things. If someone *really* wanted to call an awaited function without waiting for it, they could always use `call_deferred` or something.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here's a modification of the script from earlier, which is both much easier to read *and* write:

```gdscript
@always_await
func say(character: String, tr_string: String) -> void:
    # ...do the say thing...

func cutscene_label() -> void:
    player.position = villain.position + Vector2(0, 128)
    say("player", "ENDING_ENCOUNTER_0")
    say("player", "ENDING_ENCOUNTER_1")
    villain.animate("eyeflame")
    say("villain", "ENDING_ENCOUNTER_2")
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It's technically a workaround for writing a single keyword repeatedly, so while it technically can I'd argue the ergonomics are worth the tradeoff.

### Is there a reason why this should be core and not an add-on in the asset library?

Part of the GDScript runtime.

**Comment by IntangibleMatter** [comment]
While I see the argument for it making code less clear, I don't think this makes it any harder to debug than it otherwise is with `await`. I think the lack of clarity could be mitigated with some degree of special highlighting or something to that effect, but in long scripts where there's an `await` before almost every call to a function, and the function is never called without `await`, I think it helps with code cleanliness, as well as developer ergonomics.

**Comment by IntangibleMatter** [comment]
Hmmm... I think that the await chain would actually be a nice middle ground. It would allow the same effect (reducing redundant-feeling awaits), but would keep both readability *and* parser friendliness.

**Comment by dalexeev** [comment]
Related to / duplicate of:

* #2622
* #3901
* #6243
* #9249

**Comment by Jesusemora** [comment]
> in long scripts where there's an `await` before almost every call to a function, and the function is never called without `await`, I think it helps with code cleanliness, as well as developer ergonomics.

await used with a function tells the engine to wait for execution of the function if it has an await inside.
calling a function that has await inside without using `await` does some very neat thing:

```
func my_func() -> void:
     await get_tree().create_timer(0.5).timeout
     print("hello world")
```

```
print("1")
my_func()
print("2")
```

would print:
```
1
2
hello world
```

essentially we can run code in parallel.
using await ensures that code runs in the correct order:
```
print("1")
await my_func()
print("2")
```

would print:
```
1
hello world
2
```
and also hang execution of code for 0.5 seconds.

if you need to run multiple functions one after the other, use a `Tween`.
and await for a signal from a single function run at the end of the Tween.

```
signal chain_end
func tween_has_ended() -> void:
     chain_end.emit()
```

```
var tween : Tween = create_tween()
tween.tween_callback(self.say.bind("player", "ENDING_ENCOUNTER_0"))
tween.tween_callback(self.say.bind("player", "ENDING_ENCOUNTER_1"))
tween.tween_callback(villain.animate.bind("eyeflame"))
tween.tween_callback(self.say("villain", "ENDING_ENCOUNTER_2"))
tween.tween_callback(tween_has_ended)
await chain_end
```

just an example

**Comment by IntangibleMatter** [comment]
I would like to point out that everyone who's saying to do some sort of array of functions by using `bind()` and lambdas is missing the point of trying to *cut down* on boilerplate. While the suggestions are appreciated, the point of the proposal is to make it so *less* code can be written in situations where there doesn't need to be a keyword repeated every line for clarity.

**Comment by dalexeev** [comment]
Note that code quality is not measured by lines or characters. Sometimes boilerplate is unavoidable and not that bad. At least `await f()` is not that hard to type compared to `yield(f(), "completed")` in 3.x.

The large number of syntactic sugar in the language makes it difficult for contributors to maintain and document it, and for the community to learn and help.

`await` is somewhat of a control flow construct, just like `return`, `continue`, etc. You can't manipulate control flow as freely as you can with expressions and statements, by extracting them into separate variables/functions. Some languages ​​use macros to manipulate control flow. If GDScript supported them, you could do something like:

```gdscript
macro func SAY(character: String, tr_string: String) -> void:
    await say(character, tr_string)
```

However, macros are often criticized for various reasons. Mostly because of the lack of validation and side effects when re-evaluating an expression in the C/C++ preprocessor. But even if these problems are solved, the opinion remains that hidden control flow is bad.

I think the goal of Godot is not to solve 100% of your problems, but to solve the most difficult and common ones. For the rest, we just need to make sure that you have the ability to solve the problem yourself. Every project is unique, and it is normal that you need to create tools for it.

As mentioned above, you can use a queue if you want to avoid repeating `await`. Please elaborate on what problems this approach or just repeating `await` has? Can this proposal help not only in your project or in this particular case?

**Comment by girng** [comment]
i think the first example is more readable myself, lol. but i think i understand your motive as you probably are dealing with a lot of awaits all over the place
